File: src/utils/globalSearch.ts
import { Space, Group, Notebook } from '@/types';

export interface SearchFilters {
  spaceName: boolean;
  notebookName: boolean;
  blockTitle: boolean;
  blockContent: boolean;
}

export type GlobalSearchResultType = 'space' | 'notebook' | 'noteBlock';

export interface GlobalSearchResult {
  id: string;
  type: GlobalSearchResultType;
  matchLabels: string[];
  spaceName: string;
  notebookName?: string;
  blockTitle?: string;
  spacePath: string;
  notebookPath?: string;
  blockTitleKey?: string;
}

export interface TextSegment {
  text: string;
  highlight: boolean;
}

export const DEFAULT_SEARCH_FILTERS: SearchFilters = {
  spaceName: false,
  notebookName: true,
  blockTitle: true,
  blockContent: false,
};

export const FILTER_OPTIONS: { key: keyof SearchFilters; label: string }[] = [
  { key: 'spaceName', label: '空间名' },
  { key: 'notebookName', label: '笔记本名' },
  { key: 'blockTitle', label: '笔记块标题' },
  { key: 'blockContent', label: '笔记块内容' },
];

const MATCH_LABELS = {
  spaceName: '空间名',
  notebookName: '笔记本名',
  blockTitle: '笔记块标题',
  blockContent: '笔记块内容',
} as const;

function isGroup(item: Group | Notebook): item is Group {
  return 'children' in item;
}

function walkNotebooks(
  items: (Group | Notebook)[],
  callback: (notebook: Notebook) => boolean,
): boolean {
  for (const item of items) {
    if (isGroup(item)) {
      if (walkNotebooks(item.children, callback)) return true;
    } else if (callback(item)) {
      return true;
    }
  }
  return false;
}

export function splitHighlightSegments(text: string, query: string): TextSegment[] {
  const q = query.trim();
  if (!q) return [{ text, highlight: false }];

  const lowerText = text.toLowerCase();
  const lowerQ = q.toLowerCase();
  const segments: TextSegment[] = [];
  let lastIndex = 0;
  let index = lowerText.indexOf(lowerQ, lastIndex);

  while (index !== -1) {
    if (index > lastIndex) {
      segments.push({ text: text.slice(lastIndex, index), highlight: false });
    }
    segments.push({ text: text.slice(index, index + q.length), highlight: true });
    lastIndex = index + q.length;
    index = lowerText.indexOf(lowerQ, lastIndex);
  }

  if (lastIndex < text.length) {
    segments.push({ text: text.slice(lastIndex), highlight: false });
  }

  return segments.length > 0 ? segments : [{ text, highlight: false }];
}

export function getResultTitleParts(result: GlobalSearchResult): string[] {
  const parts = [result.spaceName];
  if (result.notebookName) parts.push(result.notebookName);
  if (result.blockTitle) parts.push(result.blockTitle);
  return parts;
}

export function performGlobalSearch(
  spaces: Space[],
  query: string,
  filters: SearchFilters,
  maxResults = 50,
): GlobalSearchResult[] {
  const q = query.trim().toLowerCase();
  if (!q) return [];

  const results: GlobalSearchResult[] = [];

  for (const space of spaces) {
    if (filters.spaceName && space.name.toLowerCase().includes(q)) {
      results.push({
        id: `space:${space.path}`,
        type: 'space',
        matchLabels: [MATCH_LABELS.spaceName],
        spaceName: space.name,
        spacePath: space.path,
      });
      if (results.length >= maxResults) return results;
    }

    walkNotebooks(space.groups, (notebook) => {
      if (filters.notebookName && notebook.name.toLowerCase().includes(q)) {
        results.push({
          id: `notebook:${notebook.path}`,
          type: 'notebook',
          matchLabels: [MATCH_LABELS.notebookName],
          spaceName: space.name,
          notebookName: notebook.name,
          spacePath: space.path,
          notebookPath: notebook.path,
        });
        if (results.length >= maxResults) return true;
      }

      for (const block of notebook.noteBlocks) {
        const matchLabels: string[] = [];
        if (filters.blockTitle && block.title.toLowerCase().includes(q)) {
          matchLabels.push(MATCH_LABELS.blockTitle);
        }
        if (filters.blockContent && block.content.toLowerCase().includes(q)) {
          matchLabels.push(MATCH_LABELS.blockContent);
        }
        if (matchLabels.length > 0) {
          results.push({
            id: `block:${notebook.path}:${block.title}`,
            type: 'noteBlock',
            matchLabels,
            spaceName: space.name,
            notebookName: notebook.name,
            blockTitle: block.title,
            spacePath: space.path,
            notebookPath: notebook.path,
            blockTitleKey: block.title,
          });
          if (results.length >= maxResults) return true;
        }
      }
      return false;
    });

    if (results.length >= maxResults) return results;
  }

  return results;
}

File: mobile/lib/core/global_search.dart
import '../l10n/l10n.dart';
import 'types.dart';

class SearchFilters {
  const SearchFilters({
    this.spaceName = false,
    this.notebookName = true,
    this.blockTitle = true,
    this.blockContent = false,
  });

  final bool spaceName;
  final bool notebookName;
  final bool blockTitle;
  final bool blockContent;

  SearchFilters copyWith({
    bool? spaceName,
    bool? notebookName,
    bool? blockTitle,
    bool? blockContent,
  }) {
    return SearchFilters(
      spaceName: spaceName ?? this.spaceName,
      notebookName: notebookName ?? this.notebookName,
      blockTitle: blockTitle ?? this.blockTitle,
      blockContent: blockContent ?? this.blockContent,
    );
  }

  bool operator [](SearchFilterKey key) {
    return switch (key) {
      SearchFilterKey.spaceName => spaceName,
      SearchFilterKey.notebookName => notebookName,
      SearchFilterKey.blockTitle => blockTitle,
      SearchFilterKey.blockContent => blockContent,
    };
  }

  SearchFilters toggled(SearchFilterKey key) {
    return switch (key) {
      SearchFilterKey.spaceName => copyWith(spaceName: !spaceName),
      SearchFilterKey.notebookName => copyWith(notebookName: !notebookName),
      SearchFilterKey.blockTitle => copyWith(blockTitle: !blockTitle),
      SearchFilterKey.blockContent => copyWith(blockContent: !blockContent),
    };
  }
}

enum SearchFilterKey { spaceName, notebookName, blockTitle, blockContent }

enum GlobalSearchResultType { space, notebook, noteBlock }

class GlobalSearchResult {
  const GlobalSearchResult({
    required this.id,
    required this.type,
    required this.matchLabels,
    required this.spaceName,
    required this.spacePath,
    this.notebookName,
    this.blockTitle,
    this.notebookPath,
    this.blockTitleKey,
  });

  final String id;
  final GlobalSearchResultType type;
  final List<String> matchLabels;
  final String spaceName;
  final String? notebookName;
  final String? blockTitle;
  final String spacePath;
  final String? notebookPath;
  final String? blockTitleKey;
}

class TextSegment {
  const TextSegment({required this.text, required this.highlight});

  final String text;
  final bool highlight;
}

const defaultSearchFilters = SearchFilters();

List<(SearchFilterKey, String)> searchFilterOptions(AppStrings s) => [
  (SearchFilterKey.spaceName, s.searchFilterSpaceName),
  (SearchFilterKey.notebookName, s.searchFilterNotebookName),
  (SearchFilterKey.blockTitle, s.searchFilterBlockTitle),
  (SearchFilterKey.blockContent, s.searchFilterBlockContent),
];

String searchFilterLabel(SearchFilterKey key, AppStrings s) {
  return switch (key) {
    SearchFilterKey.spaceName => s.searchFilterSpaceName,
    SearchFilterKey.notebookName => s.searchFilterNotebookName,
    SearchFilterKey.blockTitle => s.searchFilterBlockTitle,
    SearchFilterKey.blockContent => s.searchFilterBlockContent,
  };
}

bool _walkNotebooks(
  List<LibraryItem> items,
  bool Function(Notebook notebook) callback,
) {
  for (final item in items) {
    if (item is GroupItem) {
      if (_walkNotebooks(item.group.children, callback)) return true;
    } else if (item is NotebookItem) {
      if (callback(item.notebook)) return true;
    }
  }
  return false;
}

List<TextSegment> splitHighlightSegments(String text, String query) {
  final q = query.trim();
  if (q.isEmpty) return [TextSegment(text: text, highlight: false)];

  final lowerText = text.toLowerCase();
  final lowerQ = q.toLowerCase();
  final segments = <TextSegment>[];
  var lastIndex = 0;
  var index = lowerText.indexOf(lowerQ, lastIndex);

  while (index != -1) {
    if (index > lastIndex) {
      segments.add(
        TextSegment(text: text.substring(lastIndex, index), highlight: false),
      );
    }
    segments.add(
      TextSegment(
        text: text.substring(index, index + q.length),
        highlight: true,
      ),
    );
    lastIndex = index + q.length;
    index = lowerText.indexOf(lowerQ, lastIndex);
  }

  if (lastIndex < text.length) {
    segments.add(
      TextSegment(text: text.substring(lastIndex), highlight: false),
    );
  }

  return segments.isNotEmpty
      ? segments
      : [TextSegment(text: text, highlight: false)];
}

List<String> getResultTitleParts(GlobalSearchResult result) {
  final parts = <String>[result.spaceName];
  if (result.notebookName != null) parts.add(result.notebookName!);
  if (result.blockTitle != null) parts.add(result.blockTitle!);
  return parts;
}

List<GlobalSearchResult> performGlobalSearch(
  List<Space> spaces,
  String query,
  SearchFilters filters, {
  int maxResults = 50,
}) {
  final q = query.trim().toLowerCase();
  if (q.isEmpty) return const [];

  final s = appStrings;
  final results = <GlobalSearchResult>[];

  for (final space in spaces) {
    if (filters.spaceName && space.name.toLowerCase().contains(q)) {
      results.add(
        GlobalSearchResult(
          id: 'space:${space.path}',
          type: GlobalSearchResultType.space,
          matchLabels: [searchFilterLabel(SearchFilterKey.spaceName, s)],
          spaceName: space.name,
          spacePath: space.path,
        ),
      );
      if (results.length >= maxResults) return results;
    }

    _walkNotebooks(space.groups, (notebook) {
      if (filters.notebookName && notebook.name.toLowerCase().contains(q)) {
        results.add(
          GlobalSearchResult(
            id: 'notebook:${notebook.path}',
            type: GlobalSearchResultType.notebook,
            matchLabels: [searchFilterLabel(SearchFilterKey.notebookName, s)],
            spaceName: space.name,
            notebookName: notebook.name,
            spacePath: space.path,
            notebookPath: notebook.path,
          ),
        );
        if (results.length >= maxResults) return true;
      }

      for (final block in notebook.noteBlocks) {
        final matchLabels = <String>[];
        if (filters.blockTitle && block.title.toLowerCase().contains(q)) {
          matchLabels.add(searchFilterLabel(SearchFilterKey.blockTitle, s));
        }
        if (filters.blockContent && block.content.toLowerCase().contains(q)) {
          matchLabels.add(searchFilterLabel(SearchFilterKey.blockContent, s));
        }
        if (matchLabels.isNotEmpty) {
          results.add(
            GlobalSearchResult(
              id: 'block:${notebook.path}:${block.title}',
              type: GlobalSearchResultType.noteBlock,
              matchLabels: matchLabels,
              spaceName: space.name,
              notebookName: notebook.name,
              blockTitle: block.title,
              spacePath: space.path,
              notebookPath: notebook.path,
              blockTitleKey: block.title,
            ),
          );
          if (results.length >= maxResults) return true;
        }
      }
      return false;
    });

    if (results.length >= maxResults) return results;
  }

  return results;
}
